### ISSUE TITLE
istio-agent drops every xDS client connection after the first, blocking grpc-go upgrade beyond 1.65.0

### ISSUE THREAD

**Issue by marcel808** [istio/istio#53532]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

grpc-go post 1.65.0 introduces separate xDS client connections to istiod for each named grpc service connected to that were previously handled by a single xDS client. In our usage scenario, we follow the instructions in https://istio.io/latest/blog/2021/proxyless-grpc/#in-your-kubernetes-deployment having the grpc-agent enabled and provide settings in a bootstrap.json file. 

Once we upgrade to the grpc-go client that uses multiple clients, all gRPC requests start failing due to the istio-agent side car dropping all xDS connections after the first (e.g. clients that use gRPC service a, b and c, only the first connection gets created successfully (see log below).
We expect the istio-agent to proxy the additional xDS client connections instead of being dropped.
The end result is that we're blocked on upgrading the grpc-go client beyond 1.65.0 where the multiple xDS client connections are introduced.

```
{"level":"info","time":"2024-10-11T17:57:26.160505Z","scope":"xdsproxy","msg":"Initializing with upstream address \"istiod-1-22-4.istio-system.svc:15012\" and cluster \"Kubernetes\""}
{"level":"info","time":"2024-10-11T17:57:26.427983Z","scope":"cache","msg":"generated new workload certificate","resourceName":"default","latency":267051261,"ttl":86399572022083}
{"level":"info","time":"2024-10-11T17:57:26.428066Z","scope":"cache","msg":"Root cert has changed, start rotating root cert"}
{"level":"info","time":"2024-10-11T17:57:26.428100Z","scope":"cache","msg":"returned workload certificate from cache","ttl":86399571901173}
{"level":"info","time":"2024-10-11T17:57:26.428366Z","scope":"cache","msg":"returned workload trust anchor from cache","ttl":86399571634504}
{"level":"info","time":"2024-10-11T17:57:26.428802Z","scope":"cache","msg":"returned workload trust anchor from cache","ttl":86399571198514}
{"level":"info","time":"2024-10-11T17:57:33.536595Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":1}
{"level":"warn","time":"2024-10-11T17:57:33.808855Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T17:57:33.809498Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"info","time":"2024-10-11T17:57:33.815859Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":2}
{"level":"info","time":"2024-10-11T17:57:33.816460Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":3}
{"level":"warn","time":"2024-10-11T17:57:34.817348Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T17:57:34.818000Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}

```

### Version

```Text
$ istioctl version
client version: 1.22.2
istiod version: 1.22.4
data plane version: 1.22.4 (247 proxies)

$ kubectl version
Client Version: v1.29.7
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.8-gke.1096000
```

### Additional Information

Google GKE

Original grpc-go ticket when upgrade broke gRPC calls : https://github.com/grpc/grpc-go/issues/7691

**Comment by markdroth** [comment]
Note that this is not just an issue for grpc-go; it will be an issue for all gRPC languages as well.  This stems from the fact that as part of the xDS fallback design (see [gRFC A71](https://github.com/grpc/proposal/blob/master/A71-xds-fallback.md)), we've switched to using a separate ADS stream for each data plane channel target, and that's a cross-language design.

(At the moment, gRPC C-core will not have this problem, because our subchannel sharing mechanism means that even though we still have separate ADS streams, we'll still share the underlying connection.  But we may need to disable subchannel sharing in this case if we run into MAX_CONCURRENT_STREAMS limits.)

**Comment by keithmattix** [comment]
The overlapping streams check was added in #29811. Looks like the main concern is to make sure we clean up the correct stream during race conditions. Having multiple client grpc streams is not something we've planned for from what I can tell

**Comment by markdroth** [comment]
This seems like a fairly arbitrary limitation in istio.  Why can't the proxy handle multiple streams?

Is this a connection-level limitation (i.e., can't have more than one connection) or a stream-level limitation (can't have more than one stream, even if they're all on the same connection)?  If it's a stream-level limitation, it will apply to C-core as well.

**Comment by markdroth** [comment]
Okay.  How hard is it to fix that?

**Comment by keithmattix** [comment]
I'd want @kyessenov or @howardjohn to weigh in here; we want to make sure we teardown the correct connections in our internal structs. Multiple streams means we'd need some way to identify/differentiate between them in the proxy

**Comment by kyessenov** [comment]
I defer to @howardjohn since this is a control plane constraint. There's no problem with multiple streams on the data plane.

**Comment by easwars** [comment]
This issue is blocking folks from upgrading to the latest grpc versions. Is there an update worth sharing here? Or maybe a workaround that people can use? 

Thanks.

**Comment by collinforsyth** [comment]
Bumping this, we're also running into this issue using `grpc-go` 1.65.0

**Comment by collinforsyth** [comment]
@howardjohn I'd love to help facilitate a solution and/or workaround here.

According to [this](https://github.com/grpc/grpc-go/issues/7691#issuecomment-2411889484) comment it seems like there's a belief that this is something that is within Istio. Based on your comment, it sounds like this is something that might be risky to fix, and knowing that, do you think it's unlikely to be prioritized?

I'm curious about if we could get some reasonable next steps to fix this in either implementation, as currently it looks as though there might be a standoff between these two very excellent libraries 🙂

**Comment by howardjohn** [comment]
Its not really a bug in either project per-se but the integration between them. Istio has an XDS proxy that supports a single XDS connection, because it was designed to be used with clients which only opens a single connection (which was everyone at the time of writing).

gRPC support was then added to Istio as a super experimental thing primarily driven by one person without much adoption or maintenance around it. As, essentially, a hack around some issues that I have since forgotten the details, _one_ of the setups with gRPC is to use the XDS proxy. Its not the only one - there is `grpc-simple` that does not, its `grpc-agent` that does use it.

gRPC then changed to open multiple connections which then breaks things.

Probably the best path forward is to get gRPC working without utilizing the xDS proxy at all, rather than making the xDS proxy **substantially** more complex to support a niche case that most users are not impacted by.

**Comment by markdroth** [comment]
> Probably the best path forward is to get gRPC working without utilizing the xDS proxy at all

What would be needed to make that happen?

**Comment by howardjohn** [comment]
I don't fully remember _why_ it was required. At the very least, this would mean you need to have the gRPC to do authentication to the control plane (which just amounts to pulling a JWT from a file and setting up the CA Certs I think). You probably still want to run the agent (so not just adopting grpc-simple) so it can generate the bootstrap, serve TLS certs, etc, but just not utilize the xDS proxy.

I believe this just means modifying https://github.com/istio/istio/blob/8cd72872fe171a127ea1da74257a22e1a3160129/pkg/istio-agent/grpcxds/grpc_bootstrap.go#L166 with the appropriate configuration. Could likely be bits I am missing though

**Comment by markdroth** [comment]
Ah, right, I do remember discussing that at one point.  I think we would need changes on the gRPC side to add a call credentials type to get the JWT token from the file, and I'm not sure what would be needed for the CA cert part.  Unfortunately, I think that would be a significant chunk of work, so probably not a good approach to fix the immediate problem.

**Comment by howardjohn** [comment]
I think the CA part is already possible with `file_patcher` IIUC? https://github.com/grpc/proposal/blob/master/A29-xds-tls-security.md#file_watcher-certificate-provider.

Is there not a way to just plugin any existing PerRPCCredentials?

**Comment by markdroth** [comment]
Oh, the CA cert is already on the local filesystem?  If so, yes, I think it should be possible to use that, although the details may differ by language.

The piece that we don't have is a call credential type that uses a JWT token from a file.  But if that is the only thing we're actually missing, then it might not be too hard to add.

I assume this is using normal TLS rather than mTLS, which is why we need the JWT token to authenticate the client?

I'd like input from @ejona86 and @dfawley on the feasibility of this in gRPC Java and Go.

**Comment by markdroth** [comment]
Do we need the ability to reload the JWT token when it changes, or is it going to be constant for the life of the client?

**Comment by howardjohn** [comment]
it needs to be reloaded, its a projected volume from k8s that lasts 12hr. I think its rotated more often than that, though --- once every hour IIRC

**Comment by markdroth** [comment]
The CA cert piece will depend on [gRFC A65](https://github.com/grpc/proposal/blob/master/A65-xds-mtls-creds-in-bootstrap.md), which is implemented in C-core and I believe in Go, but not yet in Java.

**Comment by markdroth** [comment]
Even for the JWT call creds, we'd also need to define a way to configure it via the xDS bootstrap config.  So there is a bit of design work to do here -- I think this is still not going to be a super quick solution.

**Comment by markdroth** [comment]
This issue is still blocking users from upgrading gRPC.  We should not let it go stale.

**Comment by kyessenov** [comment]
@markdroth Which users is this blocking? Does grpc-go ship istio agent? Is this Istio users using proxyless?

I did a rewrite of that xds proxy about a year ago, and was not even aware of grpc usage. Even if we think it once again, I'm worried it will break again.

In general, I fully support removing xds proxy and solving the authentication issues and can help there. It would save about ~10MB per pod not to run the data plane xDS proxy and connect directly from Envoy/gRPC which is a significant saving.

**Comment by markdroth** [comment]
It's blocking Istio users that use proxyless gRPC from upgrading their version of gRPC.  And IMHO, the root cause of the problem is incorrect assumptions made by Istio's xDS proxy about how the client will behave, because there is nothing in xDS that requires the client to create only a single xDS stream.  So I think this is a clear bug in Istio that needs to be fixed.

I agree that adding the necessary JWT token support to gRPC would be a good idea, but I think that's a separate effort, and it's not something that we have any staffing to prioritize on the gRPC side.  But if someone wants to contribute the necessary design and implementation to gRPC, I would be happy to support that work.

**Comment by howardjohn** [comment]
> It's blocking Istio users that use proxyless gRPC from upgrading their version of gRPC. And IMHO, the root cause of the problem is incorrect assumptions made by Istio's xDS proxy about how the client will behave, because there is nothing in xDS that requires the client to create only a single xDS stream. So I think this is a clear bug in Istio that needs to be fixed.

Its an Envoy XDS proxy, it aims to proxy Envoy XDS -- not any arbitrary XDS client. I agree with @kyessenov that the best path forward for gRPC users _and_ Envoy users is to not use the proxy. For envoy users, making this change brings substantial unnecessary risk, while for gRPC users it leads to a subpar experience.

I understand that the design decisions of gRPC have gotten it to a state where reading a file and attaching it to a header is legitimately complex, but deciding to do something actually complicated and risky (building a multiplexing XDS proxy) is not something Istio will do. If its such a high priority, and reading the token cannot be done, perhaps gRPC can revert the change in the meantime or provide a third party proxy to add headers. Istio's proxy does a lot more in its proxy (not relevant to gRPC, but is to Envoy usage); a simple 'add header' would be easy. There is probably off the shelf proxies that could do it TBH... maybe envoy 😉

**Comment by kyessenov** [comment]
Well, it sounds like we are all on the same page. The effort to remove xDS proxy is more valuable than beefing it up to make it a generic xDS proxy, so we are not going to improve it to support grpc proxyless.

**Comment by markdroth** [comment]
We can't revert the changes on the gRPC side, since they're an inherent part of supporting xDS fallback.

I don't actually see why making Istio's proxy behave correctly is so complicated or risky.  And I think that this decision is essentially the same as saying that Istio will not support proxyless gRPC.  That's obviously your call to make, but it seems fairly short-sighted.

**Comment by kyessenov** [comment]
@markdroth You don't need an xDS proxy for proxyless. There's a negative value since you're requiring a sidecar to a gRPC app, and it's costly to run Go gRPC process with lots of protos linked in. I think gRPC needs to support a generic Kubernetes deployment and handle k8s-style authentication with mounted file CAs and JWT service account tokens out-of-the-box, without requiring Istio code.

**Comment by kyessenov** [comment]
Actually, I suspect if you can figure out how to make gRPC C++ connect to istiod xDS server, we can just take that solution verbatim into Istio as a way to skip xDS proxying to save resources. It's either that or using Envoy's gRPC client which also probably needs some features.

**Comment by markdroth** [comment]
As I said above, I agree that it would be good to add JWT support to gRPC, but we're just not staffed to prioritize that right now.  It's not that complicated to do, but someone would need to contribute the design and implementation.

**Comment by markdroth** [comment]
Are there docs somewhere on how the JWT stuff works?  Does the file contain the full JWT token, in the form that can be directly added to the authorization header, or does it need some processing first?

**Comment by markdroth** [comment]
[gRFC A97](https://github.com/grpc/proposal/blob/master/A97-xds-jwt-call-creds.md) has been merged, and I've implemented it in gRPC C-core.

If someone using gRPC C++ or Python wants to try it out and confirm that it works, that would be great.  Note that right now, you'll need to build gRPC from master and set the env var `GRPC_EXPERIMENTAL_XDS_BOOTSTRAP_CALL_CREDS=true` to enable the functionality.  Once I have confirmation that this actually works, I'll remove the env var guard, so the functionality will be available by default.

**Comment by dimpavloff** [comment]
This is still a blocker for my company in transitioning from Istio sidecars to Istio gRPC proxyless.
I have a PR open https://github.com/grpc/grpc-go/pull/8431 (A97 in Go) which I hope gets progressed soon. 

@kyessenov assuming that gets merged, what would be the work required on the Istio side?

**Comment by marcel808** [comment]
> [gRFC A97](https://github.com/grpc/proposal/blob/master/A97-xds-jwt-call-creds.md) has been merged, and I've implemented it in gRPC C-core.
> 
> If someone using gRPC C++ or Python wants to try it out and confirm that it works, that would be great. Note that right now, you'll need to build gRPC from master and set the env var `GRPC_EXPERIMENTAL_XDS_BOOTSTRAP_CALL_CREDS=true` to enable the functionality. Once I have confirmation that this actually works, I'll remove the env var guard, so the functionality will be available by default.

How about google.golang.org/grpc ?

**Comment by kyessenov** [comment]
I think you'd need to change how gRPC authenticates to istiod? You probably won't need Istio Agent next to gRPC?

**Comment by dimpavloff** [comment]
> I think you'd need to change how gRPC authenticates to istiod? You probably won't need Istio Agent next to gRPC?

Thanks @kyessenov . I'm asking because I'm not sure if this earlier comment from @howardjohn is still valid:

> You probably still want to run the agent (so not just adopting grpc-simple) so it can generate the bootstrap, serve TLS certs, etc, but just not utilize the xDS proxy.
> 
> I believe this just means modifying
> 
> [istio/pkg/istio-agent/grpcxds/grpc_bootstrap.go](https://github.com/istio/istio/blob/8cd72872fe171a127ea1da74257a22e1a3160129/pkg/istio-agent/grpcxds/grpc_bootstrap.go#L166)
> 
> Line 166 in [8cd7287](/istio/istio/commit/8cd72872fe171a127ea1da74257a22e1a3160129)
> 
>  serverURI = fmt.Sprintf("unix:///%s", opts.XdsUdsPath) 
> with the appropriate configuration. Could likely be bits I am missing though
